internal/audit/kafka/publisher.go

```diff
@@ -0,0 +1,248 @@
+	ProduceSync(context.Context, ...*kgo.Record) kgo.ProduceResults
+}
+
+type Publisher struct {
+	Client         Client
+	decisionFilter audit.DecisionLogEntryFilter
+	marshaller     recordMarshaller
+	sync           bool
+	flushTimeout   time.Duration
+}
+
+func NewPublisher(conf *Conf, decisionFilter audit.DecisionLogEntryFilter) (*Publisher, error) {
+	clientOpts := []kgo.Opt{
+		kgo.ClientID(conf.ClientID),
+		kgo.SeedBrokers(conf.Brokers...),
+		kgo.DefaultProduceTopic(conf.Topic),
+		kgo.MaxBufferedRecords(conf.MaxBufferedLogs),
+	}
+
+	if _, ok := os.LookupEnv("CERBOS_DEBUG_KAFKA"); ok {
+		clientOpts = append(clientOpts, kgo.WithLogger(
+			kzap.New(zap.L().Named("kafka"), kzap.Level(kgo.LogLevelDebug)),
+		))
+	}
+
+	ack, err := formatAck(conf.Ack)
+	if err != nil {
+		return nil, err
+	}
+	clientOpts = append(clientOpts, kgo.RequiredAcks(ack))
+	if conf.Ack != AckAll {
+		clientOpts = append(clientOpts, kgo.DisableIdempotentWrite())
+	}
+
+	client, err := kgo.NewClient(clientOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	flushTimeout, err := time.ParseDuration(conf.FlushTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Publisher{
+		Client:         client,
+		decisionFilter: decisionFilter,
+		marshaller:     recordMarshaller{Encoding: conf.Encoding},
+		sync:           conf.ProduceSync,
+		flushTimeout:   flushTimeout,
+	}, nil
+}
+
+func (p *Publisher) Close() {
+	flushCtx, flushCancel := context.WithTimeout(context.Background(), p.flushTimeout)
+	defer flushCancel()
+	// TODO: Change log.Close to return an error & then log
+	_ = p.Client.Flush(flushCtx)
+
+	p.Client.Close()
+}
+
+func (p *Publisher) Backend() string {
+	return Backend
+}
+
+func (p *Publisher) Enabled() bool {
+	return true
+}
+
+func (p *Publisher) WriteAccessLogEntry(ctx context.Context, record audit.AccessLogEntryMaker) error {
+	rec, err := record()
+	if err != nil {
+		return err
+	}
+
+	msg, err := p.marshaller.MarshalAccessLogEntry(rec)
+	if err != nil {
+		return err
+	}
+
+	return p.write(ctx, msg)
+}
+
+func (p *Publisher) WriteDecisionLogEntry(ctx context.Context, record audit.DecisionLogEntryMaker) error {
+	rec, err := record()
+	if err != nil {
+		return err
+	}
+
+	if p.decisionFilter != nil {
+		rec = p.decisionFilter(rec)
+		if rec == nil {
+			return nil
+		}
+	}
+
+	msg, err := p.marshaller.MarshalDecisionLogEntry(rec)
+	if err != nil {
+		return err
+	}
+
+	return p.write(ctx, msg)
+}
+
+func (p *Publisher) write(ctx context.Context, msg *kgo.Record) error {
+	if p.sync {
+		return p.Client.ProduceSync(ctx, msg).FirstErr()
+	}
+
+	p.Client.Produce(ctx, msg, func(r *kgo.Record, err error) {
+		if err != nil {
+			// TODO: Handle via interceptor
+			ctxzap.Extract(ctx).Warn("failed to write audit log entry", zap.Error(err))
+		}
+	})
+	return nil
+}
+
+type recordMarshaller struct {
+	Encoding string
+}
+
+func (m recordMarshaller) MarshalAccessLogEntry(rec *auditv1.AccessLogEntry) (*kgo.Record, error) {
```

Comment:
We could reuse this function for both access and decision entries by declaring it like this:

```suggestion
func (m recordMarshaller) Marshal(kind []byte, rec proto.Message) (*kgo.Record, error) {
```

Reply:
At first, I had it like this, but need access to `msg.GetCallId()` & wanted to reduce the work type switching on an interface.
That said, would simplify the code & we can worry about performance a bit later given you're taking on a new dependency to manage, so manageable code makes sense 👍🏻

Reply:
Oh I didn't think about that. Maybe declare a new interface type for it then?

```go
type auditEntry interface {
  proto.Message
  GetCallId() []byte
}
```

Reply:
Refactored things, see if you like the setup now.
The allocations have been removed on the headers & while there's likely room for improvements it felt best to land the PR before optimising much further, though I've included a benchmark to aid with that.

```
> benchstat old.txt new.txt
goos: darwin
goarch: arm64
pkg: github.com/cerbos/cerbos/internal/audit/kafka
                                                  │   old.txt   │               new.txt               │
                                                  │   sec/op    │   sec/op     vs base                │
RecordMarshaller_AccessLog/encoding_json-10         672.5n ± 0%   656.1n ± 1%   -2.45% (p=0.000 n=10)
RecordMarshaller_AccessLog/encoding_protobuf-10     163.4n ± 0%   140.9n ± 0%  -13.77% (p=0.000 n=10)
RecordMarshaller_DecisionLog/encoding_json-10       693.1n ± 3%   697.9n ± 0%        ~ (p=0.183 n=10)
RecordMarshaller_DecisionLog/encoding_protobuf-10   162.6n ± 1%   150.7n ± 1%   -7.29% (p=0.000 n=10)
geomean                                             333.6n        314.0n        -5.86%

                                                  │  old.txt   │              new.txt              │
                                                  │    B/op    │    B/op     vs base               │
RecordMarshaller_AccessLog/encoding_json-10         600.0 ± 0%   584.0 ± 0%  -2.67% (p=0.000 n=10)
RecordMarshaller_AccessLog/encoding_protobuf-10     304.0 ± 0%   288.0 ± 0%  -5.26% (p=0.000 n=10)
RecordMarshaller_DecisionLog/encoding_json-10       600.0 ± 0%   584.0 ± 0%  -2.67% (p=0.000 n=10)
RecordMarshaller_DecisionLog/encoding_protobuf-10   304.0 ± 0%   288.0 ± 0%  -5.26% (p=0.000 n=10)
geomean                                             427.1        410.1       -3.97%

                                                  │  old.txt   │              new.txt               │
                                                  │ allocs/op  │ allocs/op   vs base                │
RecordMarshaller_AccessLog/encoding_json-10         15.00 ± 0%   13.00 ± 0%  -13.33% (p=0.000 n=10)
RecordMarshaller_AccessLog/encoding_protobuf-10     6.000 ± 0%   4.000 ± 0%  -33.33% (p=0.000 n=10)
RecordMarshaller_DecisionLog/encoding_json-10       15.00 ± 0%   13.00 ± 0%  -13.33% (p=0.000 n=10)
RecordMarshaller_DecisionLog/encoding_protobuf-10   6.000 ± 0%   4.000 ± 0%  -33.33% (p=0.000 n=10)
geomean                                             9.487        7.211       -23.99%
```